tapiriik/services/TrainAsONE/trainasone.py

```diff
@@ -0,0 +1,155 @@
+                raise APIException("Unable to retrieve activity list " + str(response) + " " + response.text)
+            data = response.json()
+            allItems += data["activities"]
+            if not exhaustive or "next" not in data or data["next"] is None:
+                break
+            pageUri = TRAINASONE_SERVER_URL + data["next"]
+
+        activities = []
+        exclusions = []
+        for act in allItems:
+            try:
+                activity = self._populateActivity(act)
+            except KeyError as e:
+                exclusions.append(APIExcludeActivity("Missing key in activity data " + str(e), activity_id=act["activityId"], user_exception=UserException(UserExceptionType.Corrupt)))
+                continue
+
+            logger.debug("\tActivity s/t " + str(activity.StartTime))
+            activity.ServiceData = {"id": act["activityId"]}
+            activities.append(activity)
+        return activities, exclusions
+
+    def _populateActivity(self, rawRecord):
+        ''' Populate the 1st level of the activity object with all details required for UID from  API data '''
+        activity = UploadedActivity()
+        activity.StartTime = datetime.fromtimestamp(rawRecord["start"] / 1000)
+        activity.EndTime = activity.StartTime + timedelta(seconds=rawRecord["duration"])
+        activity.Stats.Distance = ActivityStatistic(ActivityStatisticUnit.Meters, value=rawRecord["distance"])
+        activity.GPS = rawRecord["hasGps"]
+        activity.Stationary = not rawRecord["hasGps"]
+        activity.CalculateUID()
+        return activity
+
+    def DownloadActivity(self, serviceRecord, activity):
+        activity_id = activity.ServiceData["id"]
+
+        resp = requests.get(TRAINASONE_SERVER_URL + "/api/sync/activity/tcx/" + activity_id, headers=self._apiHeaders(serviceRecord.Authorization))
+
+        try:
+            TCXIO.Parse(resp.content, activity)
+        except ValueError as e:
+            raise APIExcludeActivity("TCX parse error " + str(e), user_exception=UserException(UserExceptionType.Corrupt))
+
+        return activity
+
+    def UploadActivity(self, serviceRecord, activity):
+        # Upload the workout as a .TCX file
+        uploaddata = TCXIO.Dump(activity)
+
+        headers = self._apiHeaders(serviceRecord.Authorization)
+        headers['Content-Type'] = 'application/xml'
+        resp = requests.post(TRAINASONE_SERVER_URL + "/api/sync/activity/tcx", data=uploaddata, headers=headers)
+
+        if resp.status_code != 200:
+            raise APIException(
+                "Error uploading activity - " + str(resp.status_code),
+                block=False)
+
+        responseJson = resp.json()
+
+        if not responseJson["id"]:
+            raise APIException(
+                "Error uploading activity - " + resp.Message,
+                block=False)
+
+        activityId = responseJson["id"]
+
+        return activityId
+
+    def DeleteCachedData(self, serviceRecord):
```

Comment:
Ah - the cache was WIP. I suspect it may just be best to remove all references to it?